packages/test/test-end-to-end-tests/src/test/blobs.spec.ts

```diff
@@ -129,6 +129,29 @@ describeFullCompat("blobs", (getTestObjectProvider) => {
         assert.strictEqual(bufferToString(await blobHandle.get(), "utf-8"), testString);
     });
 
+    it("isDirty state is correct before/after blob upload", async function() {
+        const testString = "teststring";
+        const testKey = "blob";
+        const container1 = await provider.makeTestContainer(testContainerConfig);
+
+        const dataStore1 = await requestFluidObject<ITestDataObject>(container1, "default");
+
+        assert.equal(container1.isDirty, false);
+        const blob = await dataStore1._runtime.uploadBlob(stringToBuffer(testString, "utf-8"));
+        assert.equal(container1.isDirty, false);
+        dataStore1._root.set(testKey, blob);
+        assert.equal(container1.isDirty, true);
+
+        const container2 = await provider.loadTestContainer(testContainerConfig);
+        const dataStore2 = await requestFluidObject<ITestDataObject>(container2, "default");
+
+        await provider.ensureSynchronized(this.timeout() / 3);
+        assert.equal(container1.isDirty, false);
+        const blobHandle = dataStore2._root.get<IFluidHandle<ArrayBufferLike>>(testKey);
+        assert(blobHandle);
+        assert.strictEqual(bufferToString(await blobHandle.get(), "utf-8"), testString);
```

Comment:
Please ensure there is coverage for 2-3 parallel blob uploads and that system maintains correct state through these transitions and gets back to normal state after full flow.
I think you need to ensure that for online, offline and a mix of offline & online modes (i.e. losing connection halfway through second-third upload).
This is clearly should have covered even before your changes (so maybe we should ask Wes for help), but I worry these changes are going to create issues in those cases (see other comments)

Reply:
Agree. Parallel blob uploading is not covered by any unit tests afaik. Once I can define an async solution I'll add corresponding unit tests